### ISSUE TITLE
Formalize KPIs

### ISSUE THREAD

**Issue by pdehaye** [OpenDreamKit/OpenDreamKit#179]
These are listed at page 21 of 97 of the [proposal](https://github.com/OpenDreamKit/OpenDreamKit/blob/master/Proposal/proposal-www.pdf). 

They need to be formalized into separate issues, and checked with reviewers for completeness. etc. 

We also wrote in the proposal:

> We will survey mathematical departments (and relevant members of other departments during year 1 and again during year 4 to gauge the awareness of the existence and capabilities of OpenDreamKit and its components, and to collect statistical data for estimating Key Performance Indicators listed in the table. The success factor is a positive change between the two surveys. Other measures not listed in the table include activity on the relevant mailing lists and forums and the numbers of downloads (though this is hard to measure for many reasons).
> - downloads: docker images, gap/sage/singular/...
> - analytics on our web site
> - github activity, mails, ...

## Proposition of metrics for our five KPI

Aim: realistic and reasonably meaningful indicators

 => Three measurement values are needed: Month 0 when possible or Month 18 (Review), Month 36 (Review), Month 48 (Final Review)

### KPI 1: More effective collaborations between researchers #180 

- Nr of Social Media followers (Twitter)
- Number of downloads of:
    - SageMath @haraldschilly 
    - PARI/GP @videlec 
    - GAP @alex-konovalov 
    - Singular @wbhart 
- Number of installs via package managers (apt/pacman/...)
- Number of users of Sagemathcloud (via W. Stein) -> who is creating lessons, accounts for institutions etc @williamstein @haraldschilly 
- Number of Cloud-GAP @alex-konovalov 
- Webstats of ODK website /sagemath.org/Tmpnb

  => ALL ABOVE INDICATORS SHOULD BE ADJUSTED REGARDING THE INVOLVEMENT OF ODK IN IT
- Attendees to workshops/conferences (number, activity of people, domain of research, gender)

### KPI 2: Higher efficiency and creativity in research, etc. #181 

- Testimonials/surveys for qualitative indicators at the end of the project @mikecroucher 
  => This KPI is more can only be emasured through testimonies considering the nature of OpenDreamKit. Maybe the help of the end-user group is needed
- End of project: see the impact of deliverables in research (qualitative, quantitative?)

### KPI 3: Accelerated innovation in research etc. #182 

- % of tickets/commits by ODK participants to:
    - Sage
    - Gap @alex-konovalov 
- Nr of contributors to ODK components
- [Number of Jupyter kernels](https://github.com/ipython/ipython/wiki/IPython-kernels-for-other-languages) (relevant for maths? or all of them?; IIRC from discussion with @VivianePons and @defeo, all of them)
- Questions on stackoverflows and similar for ODK components: end-user stories
- User-contributed packages (Sage and Gap) @alex-konovalov 
- Number of SageMath installation on Windows (Erik's work): easy to measure, as will start during ODK @embray 

### KPI 4: Researchers able to process structured and qualitative data #183 
 All @JohnCremona 
- Number of datasets in LMFDB
- Number of components that support SCSCP
- Nr of data queries on LMFDB (and knowledge produced via MMT)
  => ALL ABOVE INDICATORS SHOULD BE ADJUSTED REGARDING THE INVOLVEMENT OF ODK IN IT

### KPI 5: Increased take-up of collaborative research and data sharing etc. #184 

- Publications stats (by altmetric?)
- Nr of publications citing ODK or one of its components (see with altmetric and openaire supports how it is possible)
- Nr of tutorial attendees / Nr of tutorials downloads
@VivianePons? Number of tutorial attendees shouls a piece of cake but downloads...?

**Comment by bpilorget** [comment]
The Coordinator and WP2 leader should take action in September to start to set up indicators

**Comment by bpilorget** [comment]
# List of realistic and meaningful indicators thought through



- Number of downloads of SageMath, Pari-Gp and other systems
- Number of install aptpacman
- Number of users of Sagemathcloud (via W. Stein) -> who is creating lessons, accounts for institutions etc
- Number of Cloud-GAP users (chen number will be available)
- Webstats of ODK website /sagemath.org/Tmpmb

- % of tickets/commits by ODK participants to software
- Nr of contributors to ODK components
- Publications stats by altmetric 
- Nr of publications citing ODK or one of its components (see with altmetric and openaire supports how it is possible)
- Questions on stackoverflows and similar for ODK components
- User-contributed packages (Sage and Gap)
- Attendees to workshops/conferences (number, activity of people, domain of research, gender)
- Jupyter kernels
- Testimonies/questionnaires for qualitative indicators at the end of the project
- Nr of tutorial attendees / Nr of turorials downloads
- Installation of Windows (Erik's work): easy measure as will start during ODK
- Number of datasets in LMFDB
- Support SCSP
- Nr of requests on LMFDV and MMT

**Comment by bpilorget** [comment]
@VivianePons @nthiery @defeo  I've summarised all the relevant KPI for us. Can you proofread them please?

**Comment by JohnCremona** [comment]
On 28 October 2016 at 15:56, bpilorget <notifications@github.com> wrote:

> List of realistic and meaningful indicators thought through
>
>    - Number of downloads of SageMath, Pari-Gp and other systems
>    - Number of install aptpacman
>    - Number of users of Sagemathcloud (via W. Stein) -> who is creating
>    lessons, accounts for institutions etc
>    - Number of Cloud-GAP users (chen number will be available)
>    -
>
>    Webstats of ODK website /sagemath.org/Tmpmb
>    -
>
>    % of tickets/commits by ODK participants to software
>    - Nr of contributors to ODK components
>    - Publications stats by altmetric
>    - Nr of publications citing ODK or one of its components (see with
>    altmetric and openaire supports how it is possible)
>    - Questions on stackoverflows and similar for ODK components
>    - User-contributed packages (Sage and Gap)
>    - Attendees to workshops/conferences (number, activity of people,
>    domain of research, gender)
>    - Jupyter kernels
>    - Testimonies/questionnaires for qualitative indicators at the end of
>    the project
>    - Nr of tutorial attendees / Nr of turorials downloads
>    - Installation of Windows (Erik's work): easy measure as will start
>    during ODK
>
>
The value of this number has nothing whatsoever to do with the success or
otherwise of the ODK project!  So please remove it.


>
>    - Number of datasets in LMFDB
>
>



>
>    - Support SCSP
>
>


>
>    - Nr of requests on LMFDB and MMT
>
>
I do not know what the latter means.  Since most LMFDB developers have
never heard of ODK, I think they would be suprised to learn that ODK was
claiming credit for increased traffic to the LMFDB website.   When we have
written new interfaces to LMFDB data then we can measure those for ODK, but
that is different.

John

**Comment by bpilorget** [comment]
@JohnCremona All these indicators are roughly thought. They need to be proofread and watched over a second time so nothing is final there.

**Comment by defeo** [comment]
@bpilorget, I just edited your list to make it more precise.

@JohnCremona, this is a list of things related to ODK that we know how to measure (fairly) easily. We now have to choose which are good KPIs (we only need a handful).

Note that the individual numbers are usually bad KPIs, but composite numbers can make sense. For example, the number of requests/data queries on LMFDB (or any other project existing before ODK) are a bad indicator. But the increase in data queries per unit of time weighted by the contribution of ODK (which we can roughly measure by % of commits) is a somewhat meaningful measure.

As for the "number of datasets" in LMFDB, yes, it's a weird number not meaning much. In the proposal we promised some measure of "increased data sharing among researchers in mathematics", though. Any better idea?

**Comment by defeo** [comment]
@bpilorget # Twitter followers / other social media?

**Comment by nthiery** [comment]
On Fri, Oct 28, 2016 at 09:28:13AM -0700, John Cremona wrote:
>    > - Installation of Windows (Erik's work): easy measure as will start
>    > during ODK
>    The value of this number has nothing whatsoever to do with the
>    success or otherwise of the ODK project! So please remove it.

The formulation was probably ambiguous: the indicator is of course not
the number of installations of the Windows operating system, but of
Sage's Windows port by Erik&co. Which is very relevant to ODK.

Cheers,
		Nicolas

**Comment by jdemeyer** [comment]
The titles of the KPI's seem very vague to me. I don't know the difference between "More effective collaborations between researchers" and "Higher efficiency and creativity in research" and "Accelerated innovation in research".

**Comment by defeo** [comment]
https://github.com/OpenDreamKit/OpenDreamKit/blame/3ad46237fc71cb6ec2fde568ea1d0f7525e2a8dc/Proposal/impact.tex#L69-L75 😄

**Comment by defeo** [comment]
go to https://webservices.opendreamkit.org/piwik for opendreamkit.org webstats

**Comment by defeo** [comment]
> Questions on stackoverflows and similar for ODK components

http://webservices.opendreamkit.org/statoverflow/ (taken once per week)

**Comment by alex-konovalov** [comment]
@defeo that's nice, thanks! GAP community on stackoverflow is mostly represented at /math.stackexchange.com: see the tag `gap` here: http://math.stackexchange.com/questions/tagged/gap. Tags on other sites e.g. http://mathoverflow.net/questions/tagged/gap and http://stackoverflow.com/questions/tagged/gap-system are less popular.

**Comment by defeo** [comment]
@alex-konovalov, done.  `gap` tag will appear with the next fetch.

**Comment by bpilorget** [comment]
@defeo Thanks a lot Luca! That looks great. Will have a look at what can be done in addition to that

**Comment by bpilorget** [comment]
@defeo Is there an ODK account created for PIWIK? It asks me for it and there is nowhere I can find how to sign in

**Comment by defeo** [comment]
@bpilorget, account data is in the private ODK repo Participants (file Analytics.md)

**Comment by bpilorget** [comment]
Hello @haraldschilly @wbhart @alex-konovalov @videlec @JohnCremona @embray @williamstein 

We will soon need the first measures of the above Key performance Indicators. They are due for the review on the 26th April and the sooner we have them the better.

Can each of you take a look at the description of this issue (#179)? Your names are written next to 1 or several indicators that I and @nthiery think you are able to measure. 
The ideal date for the measurement is 28th february 2017 (end of the 1st reporting Period) but if you can only measure the indicators at the current date, that is no problem. Feel free to choose the best location to save the data. 

Keep me posted here or by mail.
Thank you all :)

**Comment by haraldschilly** [comment]
Hi @bpilorget, uhm, I'm mentioned twice, but I don't know why you think that I can help here?

* **sagemath download numbers:** we don't have that. there are many mirrors and we don't even have solid numbers for the one we're running by the project. what I can give you is are some general indicators, like number of website visits and activities on the download page, but I don't think they're accurate.

You also say that the date of measurement is Feb 28th. Regarding download numbers that makes little sense, because one day is basically random. We definitely don't have any numbers going back to ~2005, so what's the time period you're asking about? (I assume, Feb 28th is the last day)

* **SMC: who is creating lessons, accounts for institutions.:** there are zero institutional accounts, because they simply don't exist, and there is no public list about who does create lessons. Do you really want to know all those names? I am sure that would break the privacy policy. Regarding "number of users", do you mean created accounts or active engagements? Again, a time period would be necessary.

**Comment by defeo** [comment]
Hi @haraldschilly, thanks for the reply.

> what I can give you is are some general indicators, like number of website visits and activities on the download page, but I don't think they're accurate.

Yes, that will do. We already gathered some approximate download data from sources we had access to, see https://github.com/OpenDreamKit/sagemath-ppa-stats/blob/master/index.ipynb. From many inaccurate data, we can see something more or less accurate emerge.

>  Regarding download numbers that makes little sense, because one day is basically random. 

I think Benoît is meaning cumulative stats, i.e. downloads from Adam & Eve until Feb. 28. Yearly or monthly, depending on what you have, would be good.

>  Do you really want to know all those names?

No. Realistically, # of SMC users is good enough. If you have data on active engagement, that's useful too. Again, monthly or yearly since the beginning is the best.

Thanks

**Comment by haraldschilly** [comment]
ok, let's start simple, attached are monthly website users. I don't have yearly numbers...
[Analytics sagemath.org Audience Overview 20080228-20170228.xlsx](https://github.com/OpenDreamKit/OpenDreamKit/files/865094/Analytics.sagemath.org.Audience.Overview.20080228-20170228.xlsx)

**Comment by haraldschilly** [comment]
monthly download numbers: they're only from tracking clicks on the mirror pages, which I think isn't really accurate at all, but it's at least something to hold on to. The real numbers are higher. The sum across all them is half a million, though.
[Analytics sagemath.org Top Events 20090228-20170228.xlsx](https://github.com/OpenDreamKit/OpenDreamKit/files/865173/Analytics.sagemath.org.Top.Events.20090228-20170228.xlsx)

**Comment by haraldschilly** [comment]
regarding SMC, I'll discuss with @williamstein what we want to share. we have [a stats page](https://cloud.sagemath.com/7561f68d-3d97-4530-b97e-68af2fb4ed13/raw/stats.html) and besides that, and to give you a solid benchmark number, in the year 2016 there were 81000 active users.

**Comment by bpilorget** [comment]
Thanks Harald for all the info. This year's benchmark is also the occasion to see which measure is technically possible and meaningful so it's totally fine if you can't give an answer to all questions.

**Comment by JohnCremona** [comment]
Number of datasets in LMFDB:   

50 in production database (see www.lmfdb.org/api)
84 in beta database (see beta.lmfdb.org/api)

Support SCSP

Unknown.  I could not find SCSP in Google.  I presume 0.

Nr of requests on LMFDB and MMT

On LMFDB:  a snapshot from Google analytics in www.lmfdb.org shows in the last week:
1357 sessions
885 users (split 50/50 between new and returning)
6950 page views
4m21s average session duration.

**Comment by defeo** [comment]
> Support SCSP

That's SCSCP. And I don't think LMFDB supports it ATM.

**Comment by haraldschilly** [comment]
On Thu, Mar 23, 2017 at 8:52 PM, John Cremona <notifications@github.com>
wrote:

> 50 in production database (see www.lmfdb.org/api)
> 84 in beta database (see beta.lmfdb.org/api)
>

​on top of that, the number in the brackets give you the number of
documents (aka "row") in the table.​ it would be a neat exercise to sum
them up to know how many there are in total in the entire lmfdb :-)

**Comment by JohnCremona** [comment]
On 23 March 2017 at 21:06, Harald Schilly <notifications@github.com> wrote:

> On Thu, Mar 23, 2017 at 8:52 PM, John Cremona <notifications@github.com>
> wrote:
>
> > 50 in production database (see www.lmfdb.org/api)
> > 84 in beta database (see beta.lmfdb.org/api)
> >
>
> ​on top of that, the number in the brackets give you the number of
> documents (aka "row") in the table.​ it would be a neat exercise to sum
> them up to know how many there are in total in the entire lmfdb :-)
>

For the beta database that is 82824717 individual objects.  This is
something of an overcount since it includes indices and some test
collections.  It also does not include two large collections kept
separately from the main mongo database:  (1) 2^40 class groups of
imaginary quadratic fields (2.1 terabytes); (2) 103,800,788,359
zeros of the Riemann zeta function (1.3 terabytes).

**Comment by bpilorget** [comment]
PING

**Comment by kohlhase** [comment]
> Support SCSP
> That's SCSCP. And I don't think LMFDB supports it ATM.

Actually, we are building a connection beween lmfdb and MMT and that can speak SCSCP, so we expect an SCSCP-based front-end for lmfdb (but probably not before the review meeting).

**Comment by kohlhase** [comment]
We can also add numbers on the size of the MitM ontology and the Sage/GAP exports. 
Last week (after @nthiery extended the Sage export) was 

#Categories: 297
#Axioms: 26
#Methods: 1168
#Classes 111

@Jazzpirate can you update the numbers after @pdebuyl s new generation? 
@markuspf, can you please update the GAP import so that we can count.

**Comment by pdebuyl** [comment]
Expeditive autocomplete?

**Comment by kohlhase** [comment]
> Expeditive autocomplete?

indeed, sorry I meant @pdehaye

**Comment by kohlhase** [comment]
I am not sure about the way we want to do the KPIs, do we regularly collect them? In this case, we should have a service that computes statistics as the ones above.

**Comment by bpilorget** [comment]
What was decided was to collect the KPIs at months 18, 36 and 48.

**Comment by pdehaye** [comment]
Mais bonjour à ta sœur, non? ULB physique?   :-)

**Comment by bpilorget** [comment]
@videlec Can you provide numbers for PARI-GP?
@alex-konovalov Can you provide numbers for Gap?
@wbhart or @wdecker Can you provide numbers for Singular?

(See this issue descrition within KP1)

**Comment by alex-konovalov** [comment]
For GAP, main KPI is the number of citations.
* https://www.gap-system.org/Doc/Bib/bib.html is our curated database, mainly filled in from Mathscinet. It has now about 2400 publications (100 more are in the queue to be added), and statistics is available at https://www.gap-system.org/Doc/Bib/statistics.html. This will be also shown on the poster.
* Google Scholar profile for GAP and packages (http://bit.ly/gap_citations) now has 5633 citations, including 2160 since 2012. 
* GAP Twitter account has 289 Followers: https://twitter.com/gap_system.
* Packages redistributed with GAP: 131 with GAP 4.8.7 (March 2017). To compare, it was 119 in Summer 2015 immediately prior to the start of OpenDreamKit (114 in December 2013; 99 in Summer 2012 and just 75 in December 2008).

**Comment by bpilorget** [comment]
@defeo Can you tell me again what the data on this link refers to on the KPI? https://github.com/OpenDreamKit/sagemath-ppa-stats/blob/master/index.ipynb

**Comment by bpilorget** [comment]
@alex-konovalov Is it easy for you to find how many -User-contributed packages (Sage and Gap) there are?

**Comment by bpilorget** [comment]
@alex-konovalov Same question for  % of tickets/commits by ODK participants to Gap

**Comment by alex-konovalov** [comment]
@bpilorget wrote
>@alex-konovalov Is it easy for you to find how many user-contributed packages (Sage and Gap) that ODK is working on there are?

As I wrote above, 132 with the latest GAP release. I am not sure I can answer this question for Sage.

**Comment by defeo** [comment]
> @defeo Can you tell me again what the data on this link refers to on the KPI? https://github.com/OpenDreamKit/sagemath-ppa-stats/blob/master/index.ipynb

Number of (binary) installs on Ubuntu. Each color is a new Sage
version (does not reflect "how many people have installed Sage" very
precisely).

**Comment by alex-konovalov** [comment]
@bpilorget getting the number of commits is easy. Using 
```
git shortlog -sne --since="01 Sep 2015" --before="01 Mar 2017"
```
command I get 814 commits for M1-M18 only to the MAIN repository for GAP. Of course, there is more work in other repos in https://github.com/gap-system and https://github.com/gap-packages, but I assume that we're interested in these here.

**Comment by alex-konovalov** [comment]
For number of github issues/PRs, see https://help.github.com/articles/searching-issues/. Enter a command in the search field on the top of the GitHub page, for example
```
author:alex-konovalov created:2015-09-01..2017-03-01 repo:gap-system/gap
```
This way, I've got 336 issues/Pull requests submitted by ODK member during M1-M18 only to the MAIN repository for GAP (Counted three of us in St Andrews and @dimpase, is there anyone obvious missing?)

**Comment by defeo** [comment]
Singular repo: https://github.com/Singular/Sources

**Comment by videlec** [comment]
I think that nb commits is a reasonable KPI. What about the following: for each project, provide a by month statistics of nb of commits and nb of commits by ODK member. We can make bar charts with two colors. On such graphics, we should be able to locate GAP days or PARI/GP ateliers.

I wrote [this small script](http://www.labri.fr/perso/vdelecro/git_odk_stats.sh) from which you can read my contributions to Sage

    $ ~/git_odk_stats.sh "Vincent Delecroix"
    2015-09-01 : 75
    2015-10-01 : 22
    2015-11-01 : 24
    2015-12-01 : 24
    2016-01-01 : 48
    2016-02-01 : 6
    2016-03-01 : 19
    2016-04-01 : 24
    2016-05-01 : 8
    2016-06-01 : 28
    2016-07-01 : 44
    2016-08-01 : 25
    2016-09-01 : 2
    2016-10-01 : 4
    2016-11-01 : 3
    2016-12-01 : 4
    2017-01-01 : 11
    2017-02-01 : 1
    2017-03-01 : 4

If I get the data, I volunteer to make the graphics.

**Comment by dimpase** [comment]
IMHO this is an attempt to reinvent the wheel.
https://github.com/sarahsharp/foss-heartbeat
(used by me in WP7, D7.1, see report) does all of this, graphics included,
and much more.


On Thu, Apr 27, 2017 at 9:08 AM, Vincent Delecroix <notifications@github.com

**Comment by videlec** [comment]
If it was me I will take more time learning how to use foss-heartbeat rather than writing bash scripts + make graphics with matplotlib. Moreover, I don't see how foss-heartbeat can be used to provide:
- cumulative contributions by any ODK member versus the non ODK members
- access to non github repos

**Comment by alex-konovalov** [comment]
>  think that nb commits is a reasonable KPI. 
TL;DR - I don't :). IMHO that's an evidence of work, but not a serious KPI. One person can made a 1000 of 1 line commits and another 1 commit which changes 1000 lines. If the project learns how to make better commits, squash pull requests etc. over time, this way one could say that there is now less activity because of the smaller number of commits. So, it should be analysed together with other characteristics, and that's a challenge. Also, in GAP, for example, we need to look at least at some more repositories under https://github.com/gap-system and https://github.com/gap-packages organisations, not only at the main development repository.

**Comment by pdehaye** [comment]
See #237 for rationale behind closing this issue.
